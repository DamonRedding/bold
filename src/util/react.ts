import { MutableRefObject, Ref } from 'react'

export function composeRefs<T>(...refs: Array<Ref<T>>) {
    return (value: T) => refs.forEach(ref => setRef(ref, value))
}

export function setRef<T>(ref: Ref<T>, value: any) {
    if (typeof ref === 'function') {
        ref(value)
    } else if (ref) {
        (ref as MutableRefObject<T>).current = value
    }
}

/**
 * Returns the next sibling of `element` that matches the `predicate`.
 * Returns `null` if not found.
 */
export function getNextSibling(element: Element, predicate: (sibling: Element) => boolean) {
    let current: Element = element

    while (current.nextElementSibling) {
        current = current.nextElementSibling

        if (predicate(current)) {
            return current
        }
    }

    return null
}

/**
 * Returns the previous sibling of `element` that matches the `predicate`.
 * Returns `null` if not found.
 */
export function getPreviousSibling(element: Element, predicate: (sibling: Element) => boolean) {
    let current: Element = element

    while (current.previousElementSibling) {
        current = current.previousElementSibling

        if (predicate(current)) {
            return current
        }
    }

    return null
}
