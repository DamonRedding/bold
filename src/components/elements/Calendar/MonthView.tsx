import { Interpolation } from 'emotion'
import React, { CSSProperties } from 'react'

import { Theme, useStyles } from '../../../styles'
import { getUserLocale } from '../../../util/locale'

import { createMonthMatrix } from './util'

export interface MonthViewProps {
  visibleDate?: Date
  onDayClick?(day: Date): void
  onDayHover?(day: Date): void
  renderDay?(day: Date): React.ReactNode
  renderWeekName?(firstWeekDay: Date): React.ReactNode
  createDayStyles?(day: Date, props: MonthViewProps): Interpolation
}

export function MonthView(props: MonthViewProps) {
  const { visibleDate, renderDay, renderWeekName, createDayStyles, onDayClick, onDayHover } = props
  const month = createMonthMatrix(visibleDate)
  const { classes, css } = useStyles(createStyles)

  const handleDayClick = (day: Date) => () => onDayClick(day)
  const handleDayHover = (day: Date) => () => onDayHover(day)

  return (
    <table className={classes.table}>
      <thead>
        <tr>
          {month[0].map((day, d) => (
            <th key={d}>{renderWeekName(day)}</th>
          ))}
        </tr>
      </thead>
      <tbody>
        {month.map((week, w) => (
          <tr key={w}>
            {week.map((day, d) => (
              <td key={d} data-date={day.toISOString().slice(0, 10)}>
                <span
                  className={css(classes.day, createDayStyles(day, props))}
                  onClick={handleDayClick(day)}
                  onMouseOver={handleDayHover(day)}
                  role='button'
                >
                  {renderDay(day)}
                </span>
              </td>
            ))}
          </tr>
        ))}
      </tbody>
    </table>
  )
}

MonthView.defaultProps = {
  visibleDate: new Date(),
  onDayClick: () => null,
  onDayHover: () => null,
  renderDay: day => {
    const dayFormatter = new Intl.DateTimeFormat(getUserLocale(), { day: '2-digit' })
    return dayFormatter.format(day)
  },
  renderWeekName: firstWeekDay => {
    const weekFormatter = new Intl.DateTimeFormat(getUserLocale(), { weekday: 'narrow' })
    return weekFormatter.format(firstWeekDay)
  },
  createDayStyles: () => null,
} as Partial<MonthViewProps>

export const createStyles = (theme: Theme) => ({
  table: {
    borderCollapse: 'collapse',
    textAlign: 'center',
    lineHeight: '1.5rem',
    width: '100%',

    th: {
      width: '2rem',
      padding: '0.25rem 0',
    },
  } as CSSProperties,
  day: {
    width: '2rem',
    padding: '0.25rem 0',
    display: 'inline-block',
    borderRadius: '50%',

    ':hover': {
      cursor: 'pointer',
      background: theme.pallete.surface.background,
    },
  } as CSSProperties,
  active: {
    background: theme.pallete.primary.main,
    color: theme.pallete.surface.main,
  } as CSSProperties,
})
