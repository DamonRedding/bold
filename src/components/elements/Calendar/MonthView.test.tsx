import { mount, render } from 'enzyme'
import * as React from 'react'

import { withTheme } from '../../../test'

import { MonthView } from './MonthView'

it('should render correctly', () => {
    expect(render(withTheme(
        <MonthView visibleDate={new Date('2018-10-26')} />
    ))).toMatchSnapshot()
})

it('should call onDayClick with correct date when clicked on a date', () => {
    const click = jest.fn()
    const wrapper = mount(withTheme(
        <MonthView visibleDate={new Date('2018-10-26')} onDayClick={click} />
    ))
    expect(click).not.toHaveBeenCalled()
    wrapper.find('td[data-date="2018-10-20"] span').simulate('click')
    expect(click).toHaveBeenCalledWith(new Date('2018-10-20'))
})

it('should call onDayHover with correct date when hovered on a date', () => {
    const hover = jest.fn()
    const wrapper = mount(withTheme(
        <MonthView visibleDate={new Date('2018-10-26')} onDayHover={hover} />
    ))
    expect(hover).not.toHaveBeenCalled()
    wrapper.find('td[data-date="2018-10-01"] span').simulate('mouseover')
    expect(hover).toHaveBeenCalledWith(new Date('2018-10-01'))
})

it('should use renderDay prop to render day', () => {
    const renderDay = jest.fn(() => '-')
    const wrapper = mount(withTheme(
        <MonthView visibleDate={new Date('2018-10-26')} renderDay={renderDay} />
    ))
    expect(renderDay).toHaveBeenCalledTimes(35) // Called once for each day rendered on calendar
    expect(wrapper.find('td[data-date="2018-10-01"]').text()).toEqual('-')
})

it('render week name', () => {
    const renderWeek = jest.fn(() => '-')
    const wrapper = mount(withTheme(
        <MonthView visibleDate={new Date('2018-10-26')} renderWeekName={renderWeek} />
    ))
    expect(renderWeek).toHaveBeenCalledTimes(7) // Called once for each week rendered on calendar
    expect(wrapper.find('thead').text()).toEqual('-------')
})

it('should use createDayStyles function to create styles for days', () => {
    const createDayStyles = jest.fn(() => ({ color: 'red' }))
    const wrapper = mount(withTheme(
        <MonthView visibleDate={new Date('2018-10-26')} createDayStyles={createDayStyles} />
    ))
    expect(createDayStyles).toHaveBeenCalledTimes(35) // Called once for each day rendered on calendar
    expect(wrapper.render()).toMatchSnapshot()
})
