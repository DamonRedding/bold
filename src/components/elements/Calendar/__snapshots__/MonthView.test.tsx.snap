// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
.emotion-35 {
  border-collapse: collapse;
  text-align: center;
  line-height: 1.5rem;
  width: 100%;
}

.emotion-35 th {
  width: 2rem;
  padding: 0.25rem 0;
}

.emotion-0 {
  width: 2rem;
  padding: 0.25rem 0;
  display: inline-block;
  border-radius: 50%;
}

.emotion-0:hover {
  cursor: pointer;
  background: #F2F2F7;
}

<table
  class="emotion-35"
  role="listbox"
>
  <thead>
    <tr>
      <th>
        S
      </th>
      <th>
        M
      </th>
      <th>
        T
      </th>
      <th>
        W
      </th>
      <th>
        T
      </th>
      <th>
        F
      </th>
      <th>
        S
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td
        data-date="2018-09-30"
      >
        <span
          aria-selected="false"
          class="emotion-0"
          role="option"
        >
          30
        </span>
      </td>
      <td
        data-date="2018-10-01"
      >
        <span
          aria-selected="false"
          class="emotion-0"
          role="option"
        >
          01
        </span>
      </td>
      <td
        data-date="2018-10-02"
      >
        <span
          aria-selected="false"
          class="emotion-0"
          role="option"
        >
          02
        </span>
      </td>
      <td
        data-date="2018-10-03"
      >
        <span
          aria-selected="false"
          class="emotion-0"
          role="option"
        >
          03
        </span>
      </td>
      <td
        data-date="2018-10-04"
      >
        <span
          aria-selected="false"
          class="emotion-0"
          role="option"
        >
          04
        </span>
      </td>
      <td
        data-date="2018-10-05"
      >
        <span
          aria-selected="false"
          class="emotion-0"
          role="option"
        >
          05
        </span>
      </td>
      <td
        data-date="2018-10-06"
      >
        <span
          aria-selected="false"
          class="emotion-0"
          role="option"
        >
          06
        </span>
      </td>
    </tr>
    <tr>
      <td
        data-date="2018-10-07"
      >
        <span
          aria-selected="false"
          class="emotion-0"
          role="option"
        >
          07
        </span>
      </td>
      <td
        data-date="2018-10-08"
      >
        <span
          aria-selected="false"
          class="emotion-0"
          role="option"
        >
          08
        </span>
      </td>
      <td
        data-date="2018-10-09"
      >
        <span
          aria-selected="false"
          class="emotion-0"
          role="option"
        >
          09
        </span>
      </td>
      <td
        data-date="2018-10-10"
      >
        <span
          aria-selected="false"
          class="emotion-0"
          role="option"
        >
          10
        </span>
      </td>
      <td
        data-date="2018-10-11"
      >
        <span
          aria-selected="false"
          class="emotion-0"
          role="option"
        >
          11
        </span>
      </td>
      <td
        data-date="2018-10-12"
      >
        <span
          aria-selected="false"
          class="emotion-0"
          role="option"
        >
          12
        </span>
      </td>
      <td
        data-date="2018-10-13"
      >
        <span
          aria-selected="false"
          class="emotion-0"
          role="option"
        >
          13
        </span>
      </td>
    </tr>
    <tr>
      <td
        data-date="2018-10-14"
      >
        <span
          aria-selected="false"
          class="emotion-0"
          role="option"
        >
          14
        </span>
      </td>
      <td
        data-date="2018-10-15"
      >
        <span
          aria-selected="false"
          class="emotion-0"
          role="option"
        >
          15
        </span>
      </td>
      <td
        data-date="2018-10-16"
      >
        <span
          aria-selected="false"
          class="emotion-0"
          role="option"
        >
          16
        </span>
      </td>
      <td
        data-date="2018-10-17"
      >
        <span
          aria-selected="false"
          class="emotion-0"
          role="option"
        >
          17
        </span>
      </td>
      <td
        data-date="2018-10-18"
      >
        <span
          aria-selected="false"
          class="emotion-0"
          role="option"
        >
          18
        </span>
      </td>
      <td
        data-date="2018-10-19"
      >
        <span
          aria-selected="false"
          class="emotion-0"
          role="option"
        >
          19
        </span>
      </td>
      <td
        data-date="2018-10-20"
      >
        <span
          aria-selected="false"
          class="emotion-0"
          role="option"
        >
          20
        </span>
      </td>
    </tr>
    <tr>
      <td
        data-date="2018-10-21"
      >
        <span
          aria-selected="false"
          class="emotion-0"
          role="option"
        >
          21
        </span>
      </td>
      <td
        data-date="2018-10-22"
      >
        <span
          aria-selected="false"
          class="emotion-0"
          role="option"
        >
          22
        </span>
      </td>
      <td
        data-date="2018-10-23"
      >
        <span
          aria-selected="false"
          class="emotion-0"
          role="option"
        >
          23
        </span>
      </td>
      <td
        data-date="2018-10-24"
      >
        <span
          aria-selected="false"
          class="emotion-0"
          role="option"
        >
          24
        </span>
      </td>
      <td
        data-date="2018-10-25"
      >
        <span
          aria-selected="false"
          class="emotion-0"
          role="option"
        >
          25
        </span>
      </td>
      <td
        data-date="2018-10-26"
      >
        <span
          aria-selected="false"
          class="emotion-0"
          role="option"
        >
          26
        </span>
      </td>
      <td
        data-date="2018-10-27"
      >
        <span
          aria-selected="false"
          class="emotion-0"
          role="option"
        >
          27
        </span>
      </td>
    </tr>
    <tr>
      <td
        data-date="2018-10-28"
      >
        <span
          aria-selected="false"
          class="emotion-0"
          role="option"
        >
          28
        </span>
      </td>
      <td
        data-date="2018-10-29"
      >
        <span
          aria-selected="false"
          class="emotion-0"
          role="option"
        >
          29
        </span>
      </td>
      <td
        data-date="2018-10-30"
      >
        <span
          aria-selected="false"
          class="emotion-0"
          role="option"
        >
          30
        </span>
      </td>
      <td
        data-date="2018-10-31"
      >
        <span
          aria-selected="false"
          class="emotion-0"
          role="option"
        >
          31
        </span>
      </td>
      <td
        data-date="2018-11-01"
      >
        <span
          aria-selected="false"
          class="emotion-0"
          role="option"
        >
          01
        </span>
      </td>
      <td
        data-date="2018-11-02"
      >
        <span
          aria-selected="false"
          class="emotion-0"
          role="option"
        >
          02
        </span>
      </td>
      <td
        data-date="2018-11-03"
      >
        <span
          aria-selected="false"
          class="emotion-0"
          role="option"
        >
          03
        </span>
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`should use createDayStyles function to create styles for days 1`] = `
.emotion-35 {
  border-collapse: collapse;
  text-align: center;
  line-height: 1.5rem;
  width: 100%;
}

.emotion-35 th {
  width: 2rem;
  padding: 0.25rem 0;
}

.emotion-0 {
  width: 2rem;
  padding: 0.25rem 0;
  display: inline-block;
  border-radius: 50%;
  color: red;
}

.emotion-0:hover {
  cursor: pointer;
  background: #F2F2F7;
}

<table
  class="emotion-35"
  role="listbox"
>
  <thead>
    <tr>
      <th>
        S
      </th>
      <th>
        M
      </th>
      <th>
        T
      </th>
      <th>
        W
      </th>
      <th>
        T
      </th>
      <th>
        F
      </th>
      <th>
        S
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td
        data-date="2018-09-30"
      >
        <span
          aria-selected="false"
          class="emotion-0"
          role="option"
        >
          30
        </span>
      </td>
      <td
        data-date="2018-10-01"
      >
        <span
          aria-selected="false"
          class="emotion-0"
          role="option"
        >
          01
        </span>
      </td>
      <td
        data-date="2018-10-02"
      >
        <span
          aria-selected="false"
          class="emotion-0"
          role="option"
        >
          02
        </span>
      </td>
      <td
        data-date="2018-10-03"
      >
        <span
          aria-selected="false"
          class="emotion-0"
          role="option"
        >
          03
        </span>
      </td>
      <td
        data-date="2018-10-04"
      >
        <span
          aria-selected="false"
          class="emotion-0"
          role="option"
        >
          04
        </span>
      </td>
      <td
        data-date="2018-10-05"
      >
        <span
          aria-selected="false"
          class="emotion-0"
          role="option"
        >
          05
        </span>
      </td>
      <td
        data-date="2018-10-06"
      >
        <span
          aria-selected="false"
          class="emotion-0"
          role="option"
        >
          06
        </span>
      </td>
    </tr>
    <tr>
      <td
        data-date="2018-10-07"
      >
        <span
          aria-selected="false"
          class="emotion-0"
          role="option"
        >
          07
        </span>
      </td>
      <td
        data-date="2018-10-08"
      >
        <span
          aria-selected="false"
          class="emotion-0"
          role="option"
        >
          08
        </span>
      </td>
      <td
        data-date="2018-10-09"
      >
        <span
          aria-selected="false"
          class="emotion-0"
          role="option"
        >
          09
        </span>
      </td>
      <td
        data-date="2018-10-10"
      >
        <span
          aria-selected="false"
          class="emotion-0"
          role="option"
        >
          10
        </span>
      </td>
      <td
        data-date="2018-10-11"
      >
        <span
          aria-selected="false"
          class="emotion-0"
          role="option"
        >
          11
        </span>
      </td>
      <td
        data-date="2018-10-12"
      >
        <span
          aria-selected="false"
          class="emotion-0"
          role="option"
        >
          12
        </span>
      </td>
      <td
        data-date="2018-10-13"
      >
        <span
          aria-selected="false"
          class="emotion-0"
          role="option"
        >
          13
        </span>
      </td>
    </tr>
    <tr>
      <td
        data-date="2018-10-14"
      >
        <span
          aria-selected="false"
          class="emotion-0"
          role="option"
        >
          14
        </span>
      </td>
      <td
        data-date="2018-10-15"
      >
        <span
          aria-selected="false"
          class="emotion-0"
          role="option"
        >
          15
        </span>
      </td>
      <td
        data-date="2018-10-16"
      >
        <span
          aria-selected="false"
          class="emotion-0"
          role="option"
        >
          16
        </span>
      </td>
      <td
        data-date="2018-10-17"
      >
        <span
          aria-selected="false"
          class="emotion-0"
          role="option"
        >
          17
        </span>
      </td>
      <td
        data-date="2018-10-18"
      >
        <span
          aria-selected="false"
          class="emotion-0"
          role="option"
        >
          18
        </span>
      </td>
      <td
        data-date="2018-10-19"
      >
        <span
          aria-selected="false"
          class="emotion-0"
          role="option"
        >
          19
        </span>
      </td>
      <td
        data-date="2018-10-20"
      >
        <span
          aria-selected="false"
          class="emotion-0"
          role="option"
        >
          20
        </span>
      </td>
    </tr>
    <tr>
      <td
        data-date="2018-10-21"
      >
        <span
          aria-selected="false"
          class="emotion-0"
          role="option"
        >
          21
        </span>
      </td>
      <td
        data-date="2018-10-22"
      >
        <span
          aria-selected="false"
          class="emotion-0"
          role="option"
        >
          22
        </span>
      </td>
      <td
        data-date="2018-10-23"
      >
        <span
          aria-selected="false"
          class="emotion-0"
          role="option"
        >
          23
        </span>
      </td>
      <td
        data-date="2018-10-24"
      >
        <span
          aria-selected="false"
          class="emotion-0"
          role="option"
        >
          24
        </span>
      </td>
      <td
        data-date="2018-10-25"
      >
        <span
          aria-selected="false"
          class="emotion-0"
          role="option"
        >
          25
        </span>
      </td>
      <td
        data-date="2018-10-26"
      >
        <span
          aria-selected="false"
          class="emotion-0"
          role="option"
        >
          26
        </span>
      </td>
      <td
        data-date="2018-10-27"
      >
        <span
          aria-selected="false"
          class="emotion-0"
          role="option"
        >
          27
        </span>
      </td>
    </tr>
    <tr>
      <td
        data-date="2018-10-28"
      >
        <span
          aria-selected="false"
          class="emotion-0"
          role="option"
        >
          28
        </span>
      </td>
      <td
        data-date="2018-10-29"
      >
        <span
          aria-selected="false"
          class="emotion-0"
          role="option"
        >
          29
        </span>
      </td>
      <td
        data-date="2018-10-30"
      >
        <span
          aria-selected="false"
          class="emotion-0"
          role="option"
        >
          30
        </span>
      </td>
      <td
        data-date="2018-10-31"
      >
        <span
          aria-selected="false"
          class="emotion-0"
          role="option"
        >
          31
        </span>
      </td>
      <td
        data-date="2018-11-01"
      >
        <span
          aria-selected="false"
          class="emotion-0"
          role="option"
        >
          01
        </span>
      </td>
      <td
        data-date="2018-11-02"
      >
        <span
          aria-selected="false"
          class="emotion-0"
          role="option"
        >
          02
        </span>
      </td>
      <td
        data-date="2018-11-03"
      >
        <span
          aria-selected="false"
          class="emotion-0"
          role="option"
        >
          03
        </span>
      </td>
    </tr>
  </tbody>
</table>
`;
